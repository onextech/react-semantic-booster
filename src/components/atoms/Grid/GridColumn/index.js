import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Grid as suiGrid } from 'semantic-ui-react';
import { verticalAlignFlexCssMap } from '../../../../utils/constants';
import { MediaCss } from '../../../../utils/responsive';
import { getCustomClassName, subtractObject } from '../../../../utils/helpers';


const paddedClassName = 'padded';
const containerClassName = 'container';

const StyledColumn = styled(suiGrid.Column)`
  ${MediaCss.min.sm`height: 100%;`}
  display: flex;
  font-size: 120%;
  
  // Fix for column verticalAlign with the addition of the container element
  ${({ verticalAlign }) => {
    if (verticalAlign) {
      const vaCssValue = verticalAlignFlexCssMap[verticalAlign];
      return `
        &.column.aligned.${verticalAlign} .${containerClassName} {
          height: 100%;
          display: flex;
          align-items: ${vaCssValue};
        }
      `;
    }
    return false;
  }}
  
  // background
  ${({ background }) => {
    if (background) {
      const {
        src,
        repeat,
        position,
        opacity,
        fromColor,
        toColor,
        size,
        attachment,
        color,
        overlay = true,
      } = background;
      const defaultOpacity = 0.8;
      const linearOpacity = overlay ? (opacity || defaultOpacity) : 0;
      const defaultLinearColor = 'rgba(0,0,0,' + linearOpacity + ')'; // eslint-disable-line
      const linearFrom = fromColor || defaultLinearColor;
      const linearTo = toColor || defaultLinearColor;
      return `
        &.column {
          background-color: ${color || 'black'};
          ${src && `
            background: linear-gradient(${linearFrom}, ${linearTo}), url('${src}');
            background-position: ${position || 'center'};
            background-repeat: ${repeat || 'no-repeat'};
            background-size: ${size || 'cover'};
            background-attachment: ${attachment || 'scroll'};`}
        }`;
    }
    return false;
  }}
  
  // padded
  &.${paddedClassName} .${containerClassName} {
    width: 75%;
    margin: 0 auto;
    padding: 5em 0;
    p { line-height: 1.6; } 
  }
  `;

const GridColumn = (props) => {
  // 1. Define custom props for this component
  const customProps = {
    padded: paddedClassName,
  };
  // 1.5. Define default props
  const defaultProps = {
    padded: false,
  };
  const allProps = { ...defaultProps, ...props };
  // 2. Render the custom class names
  const className = props &&
    getCustomClassName(customProps, allProps);
  // 3. Clean up custom props from main prop set to avoid prop clashing
  const semanticProps = props &&
    subtractObject({ ...defaultProps, ...customProps }, allProps);

  return (
    <StyledColumn {...semanticProps} className={className}>
      <div className={containerClassName}>
        {props.children}
      </div>
    </StyledColumn>
  );
};

GridColumn.propTypes = {
  padded: PropTypes.bool,
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.string,
    PropTypes.array,
  ]).isRequired,
};

export default GridColumn;
