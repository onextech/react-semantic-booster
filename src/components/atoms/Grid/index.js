import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Grid as suiGrid } from 'semantic-ui-react';
import GridColumn from './GridColumn';
import { verticalAlignFlexCssMap } from '../../../utils/constants';
import { getCustomClassName, subtractObject } from '../../../utils/helpers';


const attachedClassName = 'attached';
const fluidClassName = 'fluid';

const StyledGrid = styled(suiGrid)`
  &.ui.grid.inverted {
    color: white;
  }
  &.ui.grid.${attachedClassName} > .row {
    padding-bottom: initial;
  }
  &.ui.grid.${fluidClassName} > .row > .column {
    &:not(:first-of-type) {
      padding-left: 0;
    }
    &:not(:last-of-type) {
      padding-right: 0;
    }
  }
`;

const Grid = (props) => {
  // 1. Define custom props for this component
  const customProps = {
    attached: attachedClassName,
    fluid: fluidClassName,
  };
  // 1.5. Define default props
  const defaultProps = {
    attached: false,
    fluid: false,
  };
  const allProps = { ...defaultProps, ...props };
  // 2. Render the custom class names
  const className = props &&
    getCustomClassName(customProps, allProps);
  // 3. Clean up custom props from main prop set to avoid prop clashing
  const semanticProps = props &&
    subtractObject({ ...defaultProps, ...customProps }, allProps);

  return (
    <StyledGrid {...semanticProps} className={className} />
  );
};

Grid.Row = suiGrid.Row;
Grid.Column = GridColumn;

Grid.propTypes = {
  attached: PropTypes.bool,
  fluid: PropTypes.bool,
  verticalAlign: PropTypes.oneOf(Object.keys(verticalAlignFlexCssMap)),
  background: PropTypes.shape({
    src: PropTypes.string,
    repeat: PropTypes.string,
    position: PropTypes.string,
    opacity: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
    fromColor: PropTypes.string,
    toColor: PropTypes.string,
    size: PropTypes.string,
    attachment: PropTypes.string,
    color: PropTypes.string,
    overlay: PropTypes.bool,
  }),
};

export default Grid;
