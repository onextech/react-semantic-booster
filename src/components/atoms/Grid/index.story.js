import React from 'react';
import { storiesOf } from '@storybook/react';
import Grid from '.';

storiesOf('Grid', module)
  .add('Default', () => (
    <Grid />
  ));
