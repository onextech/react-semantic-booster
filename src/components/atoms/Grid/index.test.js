import React from 'react';
import renderer from 'react-test-renderer';
import Grid from '.';


describe('<Grid>', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Grid />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
