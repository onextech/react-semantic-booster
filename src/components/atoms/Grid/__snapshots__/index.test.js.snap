// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid> renders correctly 1`] = `
<div
  className="ui grid Grid__StyledGrid-hPdIDY kWsVRI"
/>
`;
